File: mood_diary.rb
# mood_diary.rb
require 'json'
require 'date'

class MoodEntry
  attr_accessor :id, :timestamp, :mood, :note

  def initialize(id, mood, note, timestamp = DateTime.now.iso8601)
    @id = id
    @timestamp = timestamp
    @mood = mood
    @note = note
  end

  def to_h
    { id: @id, timestamp: @timestamp, mood: @mood, note: @note }
  end

  def self.from_h(hash)
    MoodEntry.new(hash[:id], hash[:mood], hash[:note], hash[:timestamp])
  end
end

class MoodDiary
  attr_reader :entries

  def initialize
    @entries = []
    @next_id = 1
  end

  def add_entry(mood, note)
    raise ArgumentError, "Оценка должна быть от 1 до 10" unless (1..10).include?(mood)
    entry = MoodEntry.new(@next_id, mood, note)
    @entries << entry
    @next_id += 1
    entry
  end

  def find_entry(id)
    @entries.find { |e| e.id == id }
  end

  def delete_entry(id)
    entry = find_entry(id)
    return false unless entry
    @entries.delete(entry)
    true
  end

  def stats
    total = @entries.size
    if total == 0
      { total: 0, average: nil, min: nil, max: nil }
    else
      moods = @entries.map(&:mood)
      {
        total: total,
        average: moods.sum.to_f / total,
        min: moods.min,
        max: moods.max
      }
    end
  end

  def save_to_file(filename = "mood_data.json")
    File.write(filename, JSON.pretty_generate(@entries.map(&:to_h)))
  end

  def load_from_file(filename = "mood_data.json")
    return unless File.exist?(filename)
    data = JSON.parse(File.read(filename), symbolize_names: true)
    @entries.clear
    data.each do |item|
      entry = MoodEntry.from_h(item)
      @entries << entry
      @next_id = entry.id + 1 if entry.id >= @next_id
    end
  rescue JSON::ParserError
    puts "Ошибка чтения файла."
  end
end

def print_entry(entry)
  emoji = if entry.mood >= 7 then "😄" elsif entry.mood >= 4 then "😐" else "😞" end
  puts "#{emoji} ##{entry.id} - #{entry.timestamp} - Оценка: #{entry.mood}/10"
  puts "   Заметка: #{entry.note}"
end

def main
  diary = MoodDiary.new
  diary.load_from_file

  loop do
    puts "\n===== ДНЕВНИК НАСТРОЕНИЯ (Ruby) ====="
    puts "1. Добавить запись"
    puts "2. Показать все записи"
    puts "3. Показать статистику"
    puts "4. Удалить запись"
    puts "5. Сохранить в файл"
    puts "6. Загрузить из файла"
    puts "0. Выход"
    print "Выберите действие: "
    choice = gets.chomp

    case choice
    when "0"
      break
    when "1"
      print "Оценка настроения (от 1 до 10): "
      mood = gets.chomp.to_i
      unless (1..10).include?(mood)
        puts "Оценка должна быть от 1 до 10."
        next
      end
      print "Заметка: "
      note = gets.chomp
      entry = diary.add_entry(mood, note)
      puts "Запись добавлена с ID #{entry.id}"
    when "2"
      if diary.entries.empty?
        puts "Нет записей."
      else
        diary.entries.each { |e| print_entry(e) }
      end
    when "3"
      stats = diary.stats
      if stats[:total] == 0
        puts "Нет записей для статистики."
      else
        puts "\n=== СТАТИСТИКА ==="
        puts "Всего записей: #{stats[:total]}"
        puts "Средняя оценка: #{'%.2f' % stats[:average]}"
        puts "Минимальная оценка: #{stats[:min]}"
        puts "Максимальная оценка: #{stats[:max]}"
      end
    when "4"
      print "Введите ID записи для удаления: "
      id = gets.chomp.to_i
      if diary.delete_entry(id)
        puts "Запись удалена."
      else
        puts "Запись не найдена."
      end
    when "5"
      diary.save_to_file
      puts "Сохранено."
    when "6"
      diary.load_from_file
      puts "Загружено."
    else
      puts "Неизвестная команда."
    end
  end
end

main if __FILE__ == $0

File: mood_diary.py
# mood_diary.py
import json
from dataclasses import dataclass, asdict
from datetime import datetime
from typing import List, Optional
from pathlib import Path

@dataclass
class MoodEntry:
    id: int
    timestamp: str
    mood: int  # 1-10
    note: str

class MoodDiary:
    def __init__(self):
        self.entries: List[MoodEntry] = []
        self.next_id = 1

    def add_entry(self, mood: int, note: str) -> MoodEntry:
        if not (1 <= mood <= 10):
            raise ValueError("Оценка должна быть от 1 до 10")
        entry = MoodEntry(
            id=self.next_id,
            timestamp=datetime.now().isoformat(),
            mood=mood,
            note=note
        )
        self.entries.append(entry)
        self.next_id += 1
        return entry

    def find_entry(self, entry_id: int) -> Optional[MoodEntry]:
        return next((e for e in self.entries if e.id == entry_id), None)

    def delete_entry(self, entry_id: int) -> bool:
        entry = self.find_entry(entry_id)
        if entry:
            self.entries.remove(entry)
            return True
        return False

    def get_stats(self) -> dict:
        total = len(self.entries)
        if total == 0:
            return {"total": 0, "average": None, "min": None, "max": None}
        moods = [e.mood for e in self.entries]
        return {
            "total": total,
            "average": sum(moods) / total,
            "min": min(moods),
            "max": max(moods)
        }

    def save_to_file(self, filename: str = "mood_data.json") -> None:
        with open(filename, "w", encoding="utf-8") as f:
            json.dump([asdict(e) for e in self.entries], f, ensure_ascii=False, indent=2)

    def load_from_file(self, filename: str = "mood_data.json") -> None:
        path = Path(filename)
        if not path.exists():
            return
        with open(filename, "r", encoding="utf-8") as f:
            data = json.load(f)
            self.entries.clear()
            for item in data:
                entry = MoodEntry(
                    id=item["id"],
                    timestamp=item["timestamp"],
                    mood=item["mood"],
                    note=item["note"]
                )
                self.entries.append(entry)
                if entry.id >= self.next_id:
                    self.next_id = entry.id + 1

def print_entry(entry: MoodEntry) -> None:
    emoji = "😄" if entry.mood >= 7 else "😐" if entry.mood >= 4 else "😞"
    print(f"{emoji} #{entry.id} - {entry.timestamp} - Оценка: {entry.mood}/10")
    print(f"   Заметка: {entry.note}")

def main():
    diary = MoodDiary()
    diary.load_from_file()

    while True:
        print("\n===== ДНЕВНИК НАСТРОЕНИЯ (Python) =====")
        print("1. Добавить запись")
        print("2. Показать все записи")
        print("3. Показать статистику")
        print("4. Удалить запись")
        print("5. Сохранить в файл")
        print("6. Загрузить из файла")
        print("0. Выход")
        choice = input("Выберите действие: ").strip()

        if choice == "0":
            break
        elif choice == "1":
            try:
                mood = int(input("Оценка настроения (от 1 до 10): ").strip())
            except ValueError:
                print("Введите число.")
                continue
            if not (1 <= mood <= 10):
                print("Оценка должна быть от 1 до 10.")
                continue
            note = input("Заметка: ").strip()
            entry = diary.add_entry(mood, note)
            print(f"Запись добавлена с ID {entry.id}")
        elif choice == "2":
            if not diary.entries:
                print("Нет записей.")
            else:
                for entry in diary.entries:
                    print_entry(entry)
        elif choice == "3":
            stats = diary.get_stats()
            if stats["total"] == 0:
                print("Нет записей для статистики.")
            else:
                print("\n=== СТАТИСТИКА ===")
                print(f"Всего записей: {stats['total']}")
                print(f"Средняя оценка: {stats['average']:.2f}")
                print(f"Минимальная оценка: {stats['min']}")
                print(f"Максимальная оценка: {stats['max']}")
        elif choice == "4":
            try:
                entry_id = int(input("Введите ID записи для удаления: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            if diary.delete_entry(entry_id):
                print("Запись удалена.")
            else:
                print("Запись не найдена.")
        elif choice == "5":
            diary.save_to_file()
            print("Сохранено.")
        elif choice == "6":
            diary.load_from_file()
            print("Загружено.")
        else:
            print("Неизвестная команда.")

if __name__ == "__main__":
    main()
